### PR TITLE
feat(hubconf, build): loading a custom model with `torch.hub.load`

### PR DESCRIPTION
**Pull request by bottledmind**
I've added a possibility to load a custom YOLOX model with torch.hub from a checpotint and an exp files. Not gonna lie, it's inspired by https://github.com/ultralytics/yolov5, but I made it from scratch keeping original codestyle.
Added functionality also lets us load our own trained model without additional hussle.
I've removed some parameters from the original  `create_yolox_model` function and all the functions using it. I might add them back in the future for additional functionality, if you wish. 
Though, I doubt anyone loads a non-pretrained model with `torch.hub.load` because you've provided us with a training script. 
The `num_classes` was completely redundant.
I believe that this is a quality of life commit this repo needs.

### CODE REVIEW

**Comment by Chris-hughes10** [comment]
Hi @bottledmind, this looks interesting! Actually though, there are use cases for loading non-pretrained models from torch hub, and for that the `num_classes` argument is required, so can we please make sure that functionality is not lost?

**Comment by bottledmind** [comment]
Hi, @Chris-hughes10 
I can return the `pretrained:bool` param, but... do we really need `num_classes:int`? We kinda have this one in the exp file.
Or is it about loading a particular pretrained checkpoint? In this case I would rather add `ckpt` param for both custom and official models (like they have `weights` in `tf.keras.applications`)
I'm suggesting 
```
create_yolox_model("yolox_s", ckpt=None)  # load an official non-pretrained model
create_yolox_model("yolox_s, ckpt="object365")  # load  an official pretrained model
create_yolox_model("custom", exp="exp.py", ckpt=None, )  # load a custom non-pretrained model
create_yolox_model("custom", exp="exp.py", ckpt="ckpt.pth", )  # load a custom pretrained model
```
but the final decision is yours to make

**Comment by Chris-hughes10** [comment]
Hi @bottledmind, I find that the `num_classes` argument makes it really easy to use YOLOX in custom training scripts independently of the repo; in these cases, I don't have an exp at all. I actually have a blog post in progress about this approach as I find that it works really well.

This argument makes loading from torch hub incredibly flexible, as it works as a constructor function in a similar way to libraries like `timm`. You are probably correct that this will only be used by the minority, but it would be great you can leave it in!

Thanks!

**Comment by bottledmind** [comment]
@Chris-hughes10 
I've returned the `pretrained` and `num_classes` params. Hope I got what you want right

**Comment by Chris-hughes10** [comment]
Hi @bottledmind, thanks for the updates, looks good to me!

File: yolox/models/build.py
Comment (FateScript) [inline yolox/models/build.py:64]: `create_xxx_model` shares some overlap code, plz unify them.

**Comment by FateScript** [comment]
@bottledmind 

Plz merge `create_official_model` and `create_custom_model`  into `create_yolox_model`?

User could pass argument named `exp_file` to [`get_exp`](https://github.com/Megvii-BaseDetection/YOLOX/blob/c4298f800d/yolox/exp/build.py#L27), which fit your need.
Alos plz rename custom to `yolox_custom` to keep the variable the same style.

**Comment by bottledmind** [comment]
@FateScript 
Merged and renamed. Hope it fits your vision now

File: yolox/models/build.py
Comment (FateScript) [inline yolox/models/build.py:106]: Here is a bug, here, plz also help to fix this. The string here should be `yolov3`

File: yolox/models/build.py
Comment (FateScript) [inline yolox/models/build.py:32]: Code here break break compatibility,  consider calling like `create_yolox_model("yolox-s", False, "cpu")`

File: yolox/models/build.py
Comment (FateScript) [inline yolox/models/build.py:61]: yolox_custom and others differs in logic like `get_ckpt`, things like `load_state_dict` should be moved outside.

**Review by FateScript — CHANGES_REQUESTED** [review]

File: yolox/models/build.py
Comment (bottledmind) [reply yolox/models/build.py]: Fixed this one, but I came from the [original code](https://github.com/Megvii-BaseDetection/YOLOX/blob/c4298f800d99ddc930267d8503d49b4dc06cff48/yolox/models/build.py#L91). Wonder if it should be fixed with a separate pr.

File: yolox/models/build.py
Comment (bottledmind) [reply yolox/models/build.py]: Fixed

File: yolox/models/build.py
Comment (bottledmind) [reply yolox/models/build.py:62]: I didn't quite get what you mean by this.
At this point I'm considering a completely separate from `create_yolox_model` function, just to not overflow it with parameters. 
Also using `torch.hub.load` with keyword args (like `exp_path="exp.py", ckpt_path="ckpt.pth"`) seems wrong to me and it may be avoided with a separate function.
What do you thing about it?

Comment (FateScript) [reply yolox/models/build.py:62]: My bad, resovle this.

File: yolox/models/build.py
Comment (FateScript) [inline yolox/models/build.py:44]: a Tab indent in docstring.

**Review by FateScript — CHANGES_REQUESTED** [review]
Plz add a tab indent. @bottledmind After doing that, I will merge this PR.

File: yolox/models/build.py
Comment (bottledmind) [reply yolox/models/build.py]: Check if I've done what you wanted

**Review by FateScript — APPROVED** [review]

**Comment by FateScript** [comment]
Plz lint your code to pass the CI @bottledmind

**Comment by bottledmind** [comment]
@FateScript linted with `flake8`, no messages now

**Comment by bottledmind** [comment]
@FateScript 
Is there anything else I need to do for this PR to be merged?

**Comment by FateScript** [comment]
Thanks for your contribution @bottledmind 
LGTM.

**Pull request merged**
